### PR TITLE
Fix #355: Content rules are now compiled before any tab webview creat…

### LINKED ISSUE #355: Avoid Boot Race Conditions Regarding Content/Shield Rules
Currently content rules are [intentionally] compiled async on launch with no blocking. This can leak things that should otherwise be blocked. Tabs should not load content until all necessary rules are compiled and ready to be loaded into the webviews.

Order should be something like:

- load shield states into memory
- compile rule lists (must follow above)
- restore tabs (can happen independent of previous two steps)
- create webview[s] for related tabs (must happen after _all 3_ above steps have _completed_)
- inject compiled lists into the webviews created (must be last step)

### PR DESCRIPTION
**Pull request by jhreis**
…ion to avoid leaky shields.

## Pull Request Checklist

- [ ] My patch has gone through review and I have addressed review comments
- [ ] My patch has a standard commit message that looks like `Fix #123: This fixes the shattered coffee cup!`

- [ ] I have updated the *Unit Tests* to cover new or changed functionality
- [ ] I have updated the *UI Tests* to cover new or changed functionality
- [ ] I have marked the bug with `[needsuplift]`
- [ ] I have made sure that localizable strings use `NSLocalizableString()`

## Screenshots

If your patch includes user interface changes that you would like to suggest or that you would like UX to look at, please include them here.

## Notes for testing this patch

If useful, please leave notes for QA, explaining what this patch changes and how it can be best tested and verified.

### CODE REVIEW

File: Client/Frontend/Browser/BrowserViewController.swift [hunk 1]
```diff
@@ -391,6 +389,28 @@ class BrowserViewController: UIViewController {
         // links into the view from other apps.
         let dropInteraction = UIDropInteraction(delegate: self)
         view.addInteraction(dropInteraction)
+        
+        prepareBootData()
```
Comment (kylehickinson) [inline Client/Frontend/Browser/BrowserViewController.swift]: Can we not start this in `init` instead of in `viewDidLoad`?

Comment (jhreis) [reply Client/Frontend/Browser/BrowserViewController.swift]: Good point, will move.

**Review by iccub — CHANGES_REQUESTED** [review]
The PR crashes for me at launch.

zrzut ekranu 2018-10-25 o 16 25 54

**Comment by jhreis** [comment]
@iccub didn't retest after merge conflicts (bad on me). 

Everything should be fixed now.

**Review by kylehickinson — APPROVED** [review]

File: Client/Frontend/Browser/BrowserViewController.swift [hunk 2]
```diff
@@ -474,26 +489,24 @@ class BrowserViewController: UIViewController {
...
     override func viewWillAppear(_ animated: Bool) {
         super.viewWillAppear(animated)
 
-        if !displayedRestoreTabsAlert && crashedLastSession {
-            displayedRestoreTabsAlert = true
-            showRestoreTabsAlert()
-        } else {
-            tabManager.restoreTabs()
-        }
+        self.checkCrashRestoration()
```
Comment (iccub) [inline Client/Frontend/Browser/BrowserViewController.swift]: `self` not needed. Shouldn't swiftlint catch it btw?

File: Client/Frontend/Browser/TabManager.swift [hunk 3]
```diff
@@ -841,34 +832,16 @@ extension TabManager {
         }
     }
 
-    func restoreTabs() {
+    lazy var restoreTabs: () -> Tab = {
```
Comment (iccub) [inline Client/Frontend/Browser/TabManager.swift]: There are two `restoreTabs` now, a lazy var and a function on line 845. I would do some cleaning or renaming for more clarity.
`func restoreTabs` is only used in undoing closing tabs, I would just rename it, or remove it in a separate ticket if we are not going keep restoring closed tabs functionality in 1.7

**Review by iccub — CHANGES_REQUESTED** [review]
I did some tests and all looks good. I only suggest to do some renaming around `restoreTabs` method and/or lazy var

File: Client/Frontend/Browser/BrowserViewController.swift (hunk 2 above)
Comment (kylehickinson) [reply Client/Frontend/Browser/BrowserViewController.swift]: There's a [SwiftLint rule](https://github.com/realm/SwiftLint/blob/master/Source/SwiftLintFramework/Rules/Style/ExplicitSelfRule.swift) to catch it when you're not using it (i.e. `checkCrashRestoration`  -> `self.checkCrashRestoration`) but not the other way around?

Comment (jhreis) [reply Client/Frontend/Browser/BrowserViewController.swift]: Didn't get a warning on my end...?

**Review by iccub — APPROVED** [review]

File: Client/Frontend/Browser/BrowserViewController.swift (hunk 2 above)
Comment (kylehickinson) [reply Client/Frontend/Browser/BrowserViewController.swift]: Na I meant there's no rule for making sure you _don't_ use self explicitly.

**Pull request merged**